elastic/kibana
# [Ingest Pipelines] Add reroute processor

Closes https://github.com/elastic/kibana/issues/157481

## Summary

This PR adds the Reroute processor in Ingest Pipelines.

<img width="718" alt="Screenshot 2023-06-07 at 16 12 02" src="https://github.com/elastic/kibana/assets/59341489/f866863c-b7d5-4eb8-a56b-fd634df5665c">

**How to test**
1. Start elasticsearch with `yarn es snapshot` and kibana with `yarn start`.
2. Navigate to Stack Management -> Ingest Pipelines and start creating a new pipeline.
3. Start adding a new processor of type Reroute.
4. Verify that setting the Dataset or the Namespace fields disables the Destination field. Also, clearing up the Dataset and the Namespace fields should reenable the Destination Field.
5. Save the processor with the following parameters:
dataset: `nginx`
namespace: `default`
6. Click "Show request" and verify that the request is correct.
7. Click "Test pipeline" and add the following document:
```
[
  {
    "_index": "type-dataset-namespace",
    "_id": "id",
    "_source": {
      "foo": "bar"
    }
  }
]
```
8. Verify the output:
```
{
  "docs": [
    {
      "doc": {
        "_index": "type-nginx-default",
        "_id": "id",
        "_version": "-3",
        "_source": {
          "data_stream": {
            "namespace": "default",
            "type": "type",
            "dataset": "nginx"
          },
          "foo": "bar"
        },
        "_ingest": {
          "timestamp": "2023-06-07T15:18:33.59474Z"
        }
      }
    }
  ]
}
```
9. Edit the processor and, this time, clear up the Dataset and Namespace fields and set the Destination field to `test-destination`.
10. Verify the request.
11. Test the pipeline:
```
[
  {
    "_index": "type-dataset-namespace",
    "_id": "id",
    "_source": {
      "foo": "bar"
    }
  }
]
```
Should produce:
```
{
  "docs": [
    {
      "doc": {
        "_index": "test-destination",
        "_id": "id",
        "_version": "-3",
        "_source": {
          "foo": "bar"
        },
        "_ingest": {
          "timestamp": "2023-06-07T15:24:34.233092Z"
        }
      }
    }
  ]
}
```


### Checklist

Delete any items that are not applicable to this PR.
